File: apps/excel-service/utils/numbers-to-words-ru.go
package utils

import (
	"fmt"
	"strconv"
	"strings"
)

var onesMale = []string{
	"", "один", "два", "три", "четыре", "пять", "шесть", "семь", "восемь", "девять",
}

var onesFemale = []string{
	"", "одна", "две", "три", "четыре", "пять", "шесть", "семь", "восемь", "девять",
}

var teens = []string{
	"десять", "одиннадцать", "двенадцать", "тринадцать", "четырнадцать",
	"пятнадцать", "шестнадцать", "семнадцать", "восемнадцать", "девятнадцать",
}

var tens = []string{
	"", "", "двадцать", "тридцать", "сорок", "пятьдесят",
	"шестьдесят", "семьдесят", "восемьдесят", "девяносто",
}

var hundreds = []string{
	"", "сто", "двести", "триста", "четыреста", "пятьсот",
	"шестьсот", "семьсот", "восемьсот", "девятьсот",
}

var scales = []struct {
	one    string
	few    string
	many   string
	female bool
}{
	{},
	{one: "тысяча", few: "тысячи", many: "тысяч", female: true},
	{one: "миллион", few: "миллиона", many: "миллионов"},
	{one: "миллиард", few: "миллиарда", many: "миллиардов"},
	{one: "триллион", few: "триллиона", many: "триллионов"},
	{one: "квадриллион", few: "квадриллиона", many: "квадриллионов"},
	{one: "квинтиллион", few: "квинтиллиона", many: "квинтиллионов"},
}

func NumbersToWordsRu(value any) (string, error) {
	var n int64

	switch v := value.(type) {
	case int:
		n = int64(v)
	case int64:
		n = v
	case string:
		parsed, err := strconv.ParseInt(v, 10, 64)
		if err != nil {
			return "", err
		}
		n = parsed
	default:
		return "", fmt.Errorf("unsupported type")
	}

	if n == 0 {
		return "ноль", nil
	}

	if n < 0 {
		return "", fmt.Errorf("negative numbers are not supported")
	}

	groups := make([]int64, 0, len(scales))
	for n > 0 {
		groups = append(groups, n%1000)
		n /= 1000
	}

	if len(groups) > len(scales) {
		return "", fmt.Errorf("number is too large")
	}

	var parts []string
	for scaleIndex := len(groups) - 1; scaleIndex >= 0; scaleIndex-- {
		group := groups[scaleIndex]
		if group == 0 {
			continue
		}

		scale := scales[scaleIndex]
		parts = append(parts, convertThreeDigits(group, scale.female))
		if scaleIndex > 0 {
			parts = append(parts, plural(group, scale.one, scale.few, scale.many))
		}
	}

	return strings.Join(parts, " "), nil
}

func convertThreeDigits(n int64, female bool) string {
	var parts []string

	if n >= 100 {
		parts = append(parts, hundreds[n/100])
		n %= 100
	}

	if n >= 10 && n <= 19 {
		parts = append(parts, teens[n-10])
		return strings.Join(parts, " ")
	}

	if n >= 20 {
		parts = append(parts, tens[n/10])
		n %= 10
	}

	if n > 0 {
		if female {
			parts = append(parts, onesFemale[n])
		} else {
			parts = append(parts, onesMale[n])
		}
	}

	return strings.Join(parts, " ")
}

func plural(n int64, one, few, many string) string {
	n %= 100
	if n >= 11 && n <= 19 {
		return many
	}

	switch n % 10 {
	case 1:
		return one
	case 2, 3, 4:
		return few
	default:
		return many
	}
}

File: apps/web/src/utils/format-document-template-data/subutils/numbers-to-words-ru/numbers-to-words-ru.ts
import { capitalize } from 'es-toolkit'

import { HUNDREDS, ONES, ONES_FEMALE, TEENS, TENS } from '@/lib/constants'

const SCALES = [
  null,
  { one: 'тысяча', few: 'тысячи', many: 'тысяч', female: true },
  { one: 'миллион', few: 'миллиона', many: 'миллионов', female: false },
  { one: 'миллиард', few: 'миллиарда', many: 'миллиардов', female: false },
  { one: 'триллион', few: 'триллиона', many: 'триллионов', female: false },
  { one: 'квадриллион', few: 'квадриллиона', many: 'квадриллионов', female: false }
] as const

export function numberToWordsRu(n: number): string {
  if (!Number.isFinite(n) || !Number.isSafeInteger(n)) {
    throw new Error('number must be a finite safe integer')
  }

  if (n === 0) return 'ноль'

  if (n < 0) {
    throw new Error('negative numbers are not supported')
  }

  const groups: number[] = []
  let value = n

  while (value > 0) {
    groups.push(value % 1000)
    value = Math.floor(value / 1000)
  }

  if (groups.length > SCALES.length) {
    throw new Error('number is too large')
  }

  const result: string[] = []

  for (let scaleIndex = groups.length - 1; scaleIndex >= 0; scaleIndex--) {
    const group = groups[scaleIndex]
    if (group === 0) continue

    const scale = SCALES[scaleIndex]
    result.push(convertBelow1000(group, scale?.female ?? false))

    if (scale) {
      result.push(plural(group, scale.one, scale.few, scale.many))
    }
  }

  return capitalize(result.join(' ').trim())
}

function convertBelow1000(num: number, female: boolean = false): string {
  const words: string[] = []

  const h = Math.floor(num / 100)
  const t = Math.floor((num % 100) / 10)
  const o = num % 10

  if (h > 0) words.push(HUNDREDS[h])

  if (t === 1) {
    words.push(TEENS[o])
  } else {
    if (t > 1) words.push(TENS[t])
    if (o > 0) words.push(female ? ONES_FEMALE[o] : ONES[o])
  }

  return words.join(' ').trim()
}

function plural(n: number, one: string, few: string, many: string): string {
  const lastTwo = n % 100
  if (lastTwo >= 11 && lastTwo <= 19) return many

  const last = n % 10
  if (last === 1) return one
  if (last >= 2 && last <= 4) return few

  return many
}
